### DISCUSSION TITLE
Polar EPOCH request/merge

### DISCUSSION THREAD

**Discussion by keithbennett** [epochpic/epoch#126] [General]
In GitLab by Andrew Longman on 14 Aug 2019, 19:14  (GitLab issue GL#2080)

I have been working over the past while on adding in some new output features to 3DEPOCH suitable for my research and think that they may be of use to other people and in general for expanding the code. It is currently based around polar coordinates in which the x axis is the cylindrical axis and outputs new variables such as:

*  Radial electric, magnetic, and current denisty fields (Er, Br, Jr)

*  Azimuthal electric, magnetic, and current density fields (Et, Bt, Jt)

*  Field angular momentum in a given direction (Lx, Ly, Lz)

*  Particle angular momentum in a given direction (lx,ly,lz)

*  Total field angular momentum (Lx_total, Ly_total, Lz_total)

*  Total particle angular momentum (lx_total, ly_total, lz_total)

There are a few other diagnostics but I dont think they will be as useful in general to people. 

These diagnostics could be of use to anyone working in 3D who are studying polar dynamics of plasmas or who want to see things like absorption of angular momentum from CP lasers pulses etc. Wakefield simulations can also benefit from these outputs as they can be used for thinks like focussing forces (W = e*E_r - e*c*B_t) etc.

Let me know if this is of interest to either the community or the developers. I can pass over my code to merge, but its a little messy still and could use some optimization etc.

Thanks, Andrew

**Comment by keithbennett** [comment]
In GitLab by Andrew Longman on 14 Aug 2019, 19:15 

changed the description

**Comment by keithbennett** [comment]
In GitLab by Keith Bennett (@keithbennett) on 27 Aug 2019, 07:47 

Hi. Apologies for the late response. We are all busy at the moment preparing for conferences but your changes sound useful. Contributions are always welcome. Please do submit something and we should get time to look at your changes in a couple of weeks time.

**Comment by keithbennett** [comment]
In GitLab by Andrew Longman on 06 Sep 2019, 21:01 

Hi Keith, would you like me to post my additions to the code here, or email a merge request? Some sections of the code (in particular the total particle angular momentum per species) still needs some work from someone more familiar with the code but I think the rest works quite well.

**Comment by kwang12008** [comment]
Hi @keithbennett 
      Sorry to bother you, but I would like to know if epoch can now output the plasma angular momentum and laser angular momentum that 
 mentioned above?I look forward to your reply. Thanks a lot.
